### PR TITLE
Show names in top clients list from long-term data

### PR DESCRIPTION
**Pull request by yubiuser**
- **What does this PR aim to accomplish?:**

Replaces the IPs in the Top Clients list (in long-term data) with names if applicable. Since https://github.com/pi-hole/FTL/pull/1255 we store more data in the long-term database, e.g. the name of the client which made the query. We can use that to fetch the name when generating the top list.

Bildschirmfoto zu 2022-07-04 20-43-05

- **How does this PR accomplish the above?:**

Re-uses the code from https://github.com/pi-hole/AdminLTE/pull/2202 to get the name of the client if one is set.

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
- [x] I have read the above and my PR is ready for review. _Check this box to confirm_

### CODE REVIEW

**Comment by Bucking-Horn** [comment]
As this is about evaluating long term queries, does that cover for 
- a client of the same name showing up with different IPs over time?
- the same IP being assigned or claimed by differently named hosts over time?

**Comment by rdwebdesign** [comment]
This is using [`client_by_id` table](https://docs.pi-hole.net/database/ftl/#client_by_id).
_(I'm not sure about the second item)._

File: api_db.php [hunk 1]
```diff
@@ -158,7 +158,11 @@
 	{
 		$limit = "WHERE timestamp <= :until";
 	}
-	$stmt = $db->prepare('SELECT client,count(client) FROM queries '.$limit.' GROUP by client order by count(client) desc limit 20');
+    $dbquery= "Select CASE typeof(client) WHEN 'integer' THEN (";
+    $dbquery .= " SELECT CASE TRIM(name) WHEN '' THEN c.ip ELSE c.name END name FROM client_by_id c WHERE c.id = q.client)";
+    $dbquery .= " ELSE client END client, count(client) FROM query_storage q ".$limit." GROUP by client order by count(client) desc limit 20";
```
Comment (rdwebdesign) [inline api_db.php]: Suggestion:
Use upper case for reserved words (for consistency). Maybe use upper case for SQL functions, too (_TRIM, COUNT, TYPEOF_).

```suggestion
    $dbquery = "SELECT CASE typeof(client) WHEN 'integer' THEN (";
    $dbquery .= " SELECT CASE TRIM(name) WHEN '' THEN c.ip ELSE c.name END name FROM client_by_id c WHERE c.id = q.client)";
    $dbquery .= " ELSE client END client, count(client) FROM query_storage q ".$limit." GROUP BY client ORDER BY count(client) DESC LIMIT 20";
```

**Review by rdwebdesign — APPROVED** [review]

**Comment by yubiuser** [comment]
> a client of the same name showing up with different IPs over time?

Our reference is still the IP. If a `client` (long discussion what a client is) changes IP, this will create a new entry in the `client_by_id table` and therefore show up as different client in the top list.

> the same IP being assigned or claimed by differently named hosts over time

Again this will add a new entry in the database and show up as different client in the top list. The same happens if the client changes its name.

If the same IP/name combination is used by a different "real" client this will still be attributed to the same database client.
https://github.com/pi-hole/FTL/blob/ed35b17fe795b10006b789b7c09159a239dcd0e9/src/database/query-table.c#L169

**Pull request merged**